core/constraint-streams/src/main/java/ai/timefold/solver/constraint/streams/bavet/bi/BavetBiConcatNode.java

```diff
@@ -0,0 +1,19 @@
+package ai.timefold.solver.constraint.streams.bavet.bi;
+
+import ai.timefold.solver.constraint.streams.bavet.common.AbstractConcatNode;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.BiTuple;
+import ai.timefold.solver.constraint.streams.bavet.common.tuple.TupleLifecycle;
+
+public final class BavetBiConcatNode<A, B> extends AbstractConcatNode<BiTuple<A, B>> {
+
+    BavetBiConcatNode(TupleLifecycle<BiTuple<A, B>> nextNodesTupleLifecycle, int inputStoreIndexLeftOutTupleList,
+            int inputStoreIndexRightOutTupleList,
+            int outputStoreSize) {
+        super(nextNodesTupleLifecycle, inputStoreIndexLeftOutTupleList, inputStoreIndexRightOutTupleList, outputStoreSize);
+    }
+
+    @Override
+    protected BiTuple<A, B> getOutTuple(BiTuple<A, B> inTuple) {
+        return new BiTuple<>(inTuple.factA, inTuple.factB, outputStoreSize);
```

Comment:
Since concat does not modify the tuples, it doesn't need to create new ones.
I'm betting it will be much more efficient if it's rewritten as a pass-through node, without node creation.

Reply:
Tried that, the queue did not like it since it sees the inTuple was already inserted.

Reply:
Turn out we do need new tuples, since each node can only have 1 tuple source, and since concat has two parents, it must be a tuple source, and a tuple source must produce new tuples, not pass through existing ones.